daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala

```diff
@@ -216,13 +197,54 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
       } else this
     }
 
+    /**  Must be used to handle lookups iff [[com.daml.lf.transaction.ContractKeyUniquenessMode.byKeyOnly]] is false
+      */
     def handleLookup(lookup: Node.LookupByKey): Either[KeyInputError, State] = {
       // If the key has not yet been resolved, we use the resolution from the lookup node,
       // but this only makes sense if `activeState.keys` is updated by every node and not only by by-key nodes.
       require(!mode.byKeyOnly, "This method can only be used if all key nodes are considered")
-      handleLookupWith(lookup, lookup.result)
+      handleLookupWithInternal(lookup, lookup.result)
     }
 
+    /** Must be used to handle lookups iff [[com.daml.lf.transaction.ContractKeyUniquenessMode.byKeyOnly]] is true
+      * The second argument takes the contract key resolution to be given to the Daml interpreter instead of
+      * [[com.daml.lf.transaction.Node.LookupByKey.result]].
+      * This is because the iteration might currently be within a rollback scope
+      * that has already archived a contract with the key without a by-key operation
+      * (and for this reason the archival is not tracked in [[ContractStateMachine.ActiveLedgerState.keys]]),
+      * i.e., the lookup resolves to [[scala.None$]] but the correct key input is [[scala.Some$]] for some contract ID
+      * and this may matter after the rollback scope is left.
+      */
+    def handleLookupWith(
+        lookup: Node.LookupByKey,
+        keyInput: Option[ContractId],
+    ): Either[KeyInputError, State] = {
+      require(mode.byKeyOnly, "This method can only be used if only byKey nodes are considered")
+      handleLookupWithInternal(lookup, keyInput)
+    }
+
+    private def handleLookupWithInternal(
```

Comment:
Something got messed up. There is another method with the same name below.

Reply:
thx fixed it up